### PR TITLE
Feature/page/dashboard

### PR DESCRIPTION
**Pull request by picarro-sfauque**
Initial working full-stack application. Not yet connected to datapool. 

Tickets:
[typescript for some components](https://picarro.atlassian.net/browse/SAV-61)
[fastAPI learning and implementation BE](https://picarro.atlassian.net/browse/SAV-66)
[responsive FE](https://picarro.atlassian.net/browse/SAV-73) 
[datetime rangeSlider FE -> BE](https://picarro.atlassian.net/browse/SAV-74)
[datetime rangeSlider FE](https://picarro.atlassian.net/browse/SAV-71)
[FE tests](https://picarro.atlassian.net/browse/SAV-72)
[Migrate Repo](https://picarro.atlassian.net/browse/SAV-84)

Ask me for data files.
For the frontend you will have to put the `treatedDataTable.json` file in the `semi-applicaiton/frontend/src/data/` folder. 

Install Node if it is not installed https://nodejs.org/en/download/

navigate to 'semi-application/frontend' and run:

```bash
npm install
```

to install all dependencies and to start the application run

and then run

```bash
npm start
```

A webbrowser window will open once the application is ready. In the terminal run `ctrl-c` to close the server.

You will need to place the `Experiment-DefinitionGF.json` in the `semi-application/backend/src/middleware` folder and `treatedDataTabe.json` will have to go to the `semi-application/backend/src/dataframe` folder. This is until the data pool is implemented. 

To get it started, navigate to 'semi-application/backend' and run:

```bash
make develop
```

```bash
source venv/bin/activate
```

and then run

```bash
uvicorn main:app --reload
```

You will be notified with `Application startup complete` once the application is ready.

### CODE REVIEW

**Review by picarro-abardales — CHANGES_REQUESTED** [review]
Could you update your backend python app's requirements.txt file? I tried running the backend (`python src/main.py`) but got this:

```
~> python src/main.py 
Traceback (most recent call last):
  File "/home/alex/code/semi-application/backend/src/main.py", line 1, in <module>
    from fastapi.middleware.cors import CORSMiddleware
ModuleNotFoundError: No module named 'fastapi'
```

And when I looked at the requirements.txt file it seemed pretty bare.

Do you think I should be able to run the application on my laptop? If so, I'll re-review when this is fixed up a bit and it can run.

**Comment by picarro-abardales** [comment]
Also, for the front-end, I got this error after `npm install && npm run`:

```
webpack 5.70.0 compiled with 2 errors and 1 warning in 45836 ms
ERROR in src/components/Grid.tsx:4:30
TS2307: Cannot find module './BasicPlotScatter' or its corresponding type declarations.
    2 | import './Grid.css';
    3 | import AllenDevPlot from './AllenDevPlot';
  > 4 | import BasicPlotScatter from './BasicPlotScatter';
      |                              ^^^^^^^^^^^^^^^^^^^^
    5 |
    6 | const Grid: React.FC = () => {
    7 |   return (
```

There was a more verbose output that is too large to paste here but it seems to point to the fact that the server was unable to resolve a `treatedDataTable.json` file.

**Comment by picarro-sfauque** [comment]
Thanks Alex, I had forgot to prune that area of the frontend code. I'll do that now.

**Comment by picarro-sfauque** [comment]
@picarro-abardales commits have been submitted. You will need to grab the `treatedDataTable.json` file from me and place the file in the `backend/src/dataframe` folder and the `Experiment_DefinitionGF.json` file will have to go into the `backend/src/middleware` folder. These are both temporary until the data pool is connected. 

Also the command to run the React server from the `frontend` directory is `npm start`.

Thank you!

**Comment by picarro-abardales** [comment]
What is the deployment going to look like for this app? Are both the front end and back end services going to be run inside a container? The deployment strategy will affect how we handle things like paths (e.g. "experiment definition JSON") in the backend server, for example.

**Review by picarro-abardales — APPROVED** [review]
Thanks for updating the requirements.txt file.

I added a few suggestions but I'm not strongly opinionated. If I understand the PR description correctly, much of this is for demo purposes and some things will be changed out in the future.

I like how you've paired the routes with data modeling via Pydantic. This is a solid pattern and will pay dividends in the future.

File: backend/src/middleware/plot_by_port.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    instrument = doc["CAMPAIGN"]["instrument"]
+
+    # Convert the timezone string into a timezone class
+    tz = pytz.timezone(time_zone)
+
+    # Get all experiments from the EXPERIMENTS OBJECT
+    expts = doc["EXPERIMENTS"]
+
+    # Grab the RoadWork experiment dictionary from the EXPERIMENTS object
+    expt = expts["RoadWork"]
```
[the remark this answers is not in the document]
Comment (picarro-sfauque) [reply backend/src/middleware/plot_by_port.py:40]: > What is the deployment going to look like for this app? Are both the front end and back end services going to be run inside a container? The deployment strategy will affect how we handle things like paths (e.g. "experiment definition JSON") in the backend server, for example.

The app will be run inside of a container. 

> What happens when the JSON document does not exist or has a different structure? There are no guarantees that the `CAMPAIGN` key will exist, for example. Therefore, in a future PR, consider adding some schema validation to this file using a library like Pydantic. Or at least add a config parsing function.


- There will be a validator to ensure that an Experiment-Defintion.JSON document is selected before the analysis can start. 
- The current iteration will allow the user to completely define the Exp-Def the way they want (from a template in a code editor). A later version would comprise of a prefilled form that can be selected and a copy of the form could be modified and saved. I'll be sure to add schema validation with Pydantic, thanks for the tip.

**Pull request merged**